[57258] Fix issues found during review

cbliard opened the pull request:
# Ticket
https://community.openproject.org/wp/57258

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What are you trying to accomplish?

Two issues have been found during review:

- [x] Derivation hints should be displayed in status-based mode
- [x] Hints when remaining work is increased/decreased like work should be reworked

This PR addresses them.

## Screenshots

**Derivation hints should be displayed in status-based mode**

Before:

![image](https://github.com/user-attachments/assets/68c23a73-3e6f-4a08-a5f5-8e08b4f6a57e)

After:

![image](https://github.com/user-attachments/assets/7b4f5a63-cc6e-4364-85a8-dfc0089709a1)

**Hints when remaining work is increased/decreased like work should be reworked**

The amount by which it is changed is indicated in the hint text.

Before:

![Aug-29-2024_11-33-42](https://github.com/user-attachments/assets/eb34da92-3df5-4435-9f1e-f8888f3146f1)

After: 

![new behavior with amount](https://github.com/user-attachments/assets/a6ffadbe-ed1a-497d-af94-ca82dfe64315)


# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

**Derivation hints should be displayed in status-based mode**

Call the right method to set the caption from the calculated hint.
Then some hints were displayed inappropriately so they had to be fixed.

**Hints when remaining work is increased/decreased like work should be reworked**

The delta is added to the derived hint in the work package. A `WorkPackage::ProgressHint` class is introduced to manage the logic of converting the keys and the delta to a translated message with the duration converted into hours or hours and days depending on the configuration.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
[57258] Display derivation hints in status-based mode

https://community.openproject.org/wp/57258

It was calculated but not displayed. This is now covered by a feature
test.

In the progress popover, when the input is a select, the event being
listened to to render a preview is `change` instead of `input. It's
because cuprite does not emit the `input` event when selecting an option
in a select input. It only emits `change`. `ProgressEditField` has been
updated to correctly focus and change the value of a select (for
status selection inside the popover).

Also to allow `ProgressEditField#expect_modal_field_value` to work
regardless of the field being disabled or readonly, the disabled value
default to `:all` which matches both disabled and enabled inputs, and
the `readonly` parameter has been removed. To check for it, call
`ProgressEditField#expect_modal_field_read_only` directly instead.

cbliard committed:
[57258] Indicate delta in remaining work derived message

When work is increased or decreased and remaining work is also increased
or decreased by the same amount, indicate the delta in the hint message
to reduce user confusion.

See comments 12 to 18 of work package 57258 for more details.

cbliard committed:
Change i18n key to prevent inconsistent translation issues

aaron-contreras commented on app/forms/work_packages/progress_form.rb:
```diff
@@ -199,9 +197,10 @@ def as_percent(value)
     data = { "work-packages--progress--preview-progress-target": "progressInput",
              "work-packages--progress--touched-field-marker-target": "progressInput",
-             action: "input->work-packages--progress--touched-field-marker#markFieldAsTouched" }
+             action: "#{action}->work-packages--progress--touched-field-marker#markFieldAsTouched" }
```
Now that I think of it, we could even omit the action and it would use the default event for each input type. Meaning for select fields it would use "change", for input fields would use "input" and we don't need to specify this. But we could also argue that being explicit about the event is probably best.
```suggestion
             action: "work-packages--progress--touched-field-marker#markFieldAsTouched" }
```

cbliard replied:
Good catch! I tried it and it works. I'll commit it.

cbliard replied:
commited

aaron-contreras commented on frontend/src/stimulus/controllers/dynamic/work-packages/progress/preview-progress.controller.ts:
```diff
@@ -61,7 +61,11 @@ export default class PreviewProgressController extends Controller {
 
     this.progressInputTargets.forEach((target) => {
-      target.addEventListener('input', this.debouncedPreview);
+      if (target.tagName.toLowerCase() === 'select') {
+        target.addEventListener('change', this.debouncedPreview);
+      } else {
+        target.addEventListener('input', this.debouncedPreview);
+      }
```
Same comment as above - whatever you see more fit

cbliard replied:
I don't get it. The first argument is mandatory. How can I make both identical?

aaron-contreras commented on spec/support/edit_fields/progress_edit_field.rb:
```diff
@@ -87,18 +87,31 @@ def clear
     return if focused?
 
     input_element.click
+    input_element.click if status_field? # to close the dropdown
```
Clever

aaron-contreras reviewed (approved):
All looks good! Left a question open to your decision. Merge if disregarded :)

cbliard committed:
Simplify action triggering

When not specified, the default event for each input type is used. So it
will be `change` for select fields, and `input` for input fields.

Co-authored-by: Aaron Contreras <aaronlcaq@gmail.com>
